File: archlucid-ui/src/lib/audit-event-presentation.ts
/**
 * Mirrors {@link ArchLucid.Contracts.Audit.AuditEventPresentation} and {@link ReviewAuditLifecycleStage} so audit /
 * activity timelines stay aligned with server Contracts without duplicating ad hoc strings in the UI layer.
 */

export const ReviewAuditLifecycleStage = {
  ReviewStarted: 0,
  ContextCaptured: 1,
  GraphCreated: 2,
  FindingsCaptured: 3,
  ManifestFinalized: 4,
  ArtifactsBundled: 5,
  GovernanceHandoff: 6,
  Other: 99,
} as const;

export type ReviewAuditLifecycleStageValue =
  (typeof ReviewAuditLifecycleStage)[keyof typeof ReviewAuditLifecycleStage];

/** Buyer-facing title for durable audit type codes (unknown codes are humanized like the C# helper). */
export function auditEventFriendlyTitle(eventType: string): string {
  const key = eventType.trim();

  if (key.length === 0) {
    return "Event";
  }

  switch (key) {
    case "RunStarted":
      return "Review started";

    case "RunSubmitted":
      return "Review submitted";

    case "RunCompleted":
      return "Review completed";

    case "Request.Created":
      return "Architecture request captured";

    case "ManifestGenerated":
      return "Manifest generated";

    case "ManifestFinalized":
      return "Manifest finalized";

    case "ManifestViewed":
      return "Manifest viewed";

    case "ReviewTrailAccessed":
      return "Review trail accessed";

    case "ProvenanceAccessed":
      return "Evidence graph accessed";

    case "FindingsListAccessed":
      return "Findings reviewed";

    case "FindingsSnapshotSealed":
      return "Findings captured";

    case "ArtifactsGenerated":
      return "Artifacts bundled";

    case "GovernanceApprovalRequested":
      return "Governance approval requested";

    default:
      return humanizeUnknownAuditEventType(key);
  }
}

/** Coarse lifecycle bucket for grouping Activity / Audit timelines (matches Contracts enum ordinals). */
export function auditEventLifecycleStage(eventType: string): ReviewAuditLifecycleStageValue {
  const key = eventType.trim();

  switch (key) {
    case "RunStarted":
    case "RunSubmitted":
      return ReviewAuditLifecycleStage.ReviewStarted;

    case "Request.Created":
      return ReviewAuditLifecycleStage.ContextCaptured;

    case "ReviewTrailAccessed":
    case "ProvenanceAccessed":
      return ReviewAuditLifecycleStage.GraphCreated;

    case "FindingsListAccessed":
    case "FindingsSnapshotSealed":
      return ReviewAuditLifecycleStage.FindingsCaptured;

    case "ManifestGenerated":
    case "ManifestFinalized":
    case "ManifestViewed":
      return ReviewAuditLifecycleStage.ManifestFinalized;

    case "ArtifactsGenerated":
      return ReviewAuditLifecycleStage.ArtifactsBundled;

    case "GovernanceApprovalRequested":
      return ReviewAuditLifecycleStage.GovernanceHandoff;

    default:
      return ReviewAuditLifecycleStage.Other;
  }
}

/** Humanize unknown audit codes — mirrors Contracts `HumanizeUnknown` (last dotted segment, word-split, title case). */
export function humanizeUnknownAuditEventType(raw: string): string {
  const parts = raw.split(".").filter((p) => p.trim().length > 0);
  const last = parts.length > 0 ? parts[parts.length - 1] : raw;
  const words = last.replace(/-/g, " ").replace(/_/g, " ").split(/\s+/).filter((w) => w.length > 0);

  if (words.length === 0) {
    return raw;
  }

  return words
    .map((w) => (w.length === 0 ? "" : w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()))
    .join(" ");
}

File: ArchLucid.Contracts/Audit/AuditEventPresentation.cs
namespace ArchLucid.Contracts.Audit;

/// <summary>
///     Stable buyer-facing titles + lifecycle grouping for durable audit event type strings (mirror literals from
///     <c>ArchLucid.Core.Audit.AuditEventTypes</c> where applicable).
/// </summary>
public static class AuditEventPresentation
{
    /// <summary>
    ///     Returns a reviewer-facing title suitable for sponsor timelines — falls back to the trimmed raw code when unknown.
    /// </summary>
    public static string FriendlyTitle(string eventType)
    {
        string key = eventType.Trim();

        if (key.Length == 0)
        {
            return "Event";
        }

        return key switch
        {
            "RunStarted" => "Review started",

            "RunSubmitted" => "Review submitted",

            "RunCompleted" => "Review completed",

            "Request.Created" => "Architecture request captured",

            "ManifestGenerated" => "Manifest generated",

            "ManifestFinalized" => "Manifest finalized",

            "ManifestViewed" => "Manifest viewed",

            "ReviewTrailAccessed" => "Review trail accessed",

            "ProvenanceAccessed" => "Evidence graph accessed",

            "FindingsListAccessed" => "Findings reviewed",

            "FindingsSnapshotSealed" => "Findings captured",

            "ArtifactsGenerated" => "Artifacts bundled",

            "GovernanceApprovalRequested" => "Governance approval requested",

            _ => HumanizeUnknown(key),
        };
    }

    /// <summary>
    ///     Maps known audit codes into coarse lifecycle stages for grouping Activity vs Audit surfaces consistently.
    /// </summary>
    public static ReviewAuditLifecycleStage LifecycleStage(string eventType)
    {
        string key = eventType.Trim();

        return key switch
        {
            "RunStarted" or "RunSubmitted" => ReviewAuditLifecycleStage.ReviewStarted,

            "Request.Created" => ReviewAuditLifecycleStage.ContextCaptured,

            "ReviewTrailAccessed" or "ProvenanceAccessed" => ReviewAuditLifecycleStage.GraphCreated,

            "FindingsListAccessed" or "FindingsSnapshotSealed" => ReviewAuditLifecycleStage.FindingsCaptured,

            "ManifestGenerated" or "ManifestFinalized" or "ManifestViewed" => ReviewAuditLifecycleStage.ManifestFinalized,

            "ArtifactsGenerated" => ReviewAuditLifecycleStage.ArtifactsBundled,

            "GovernanceApprovalRequested" => ReviewAuditLifecycleStage.GovernanceHandoff,

            _ => ReviewAuditLifecycleStage.Other,
        };
    }

    private static string HumanizeUnknown(string raw)
    {
        string[] parts = raw.Split('.', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        string last = parts.Length > 0 ? parts[^1] : raw;

        string[] words = last
            .Replace('-', ' ')
            .Replace('_', ' ')
            .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        if (words.Length == 0)
        {
            return raw;
        }

        IEnumerable<string> titled = words.Select(static w =>
            w.Length == 0 ? string.Empty : char.ToUpperInvariant(w[0]) + (w.Length > 1 ? w[1..].ToLowerInvariant() : string.Empty));

        return string.Join(' ', titled);
    }
}
